### PR TITLE
Include Codecov in CI

### PR DESCRIPTION
**Pull request by justsmth**
### Issues:
N/A

### Description of changes: 
* Generates coverage data using `gcov` and uploads it to Codecov.

### Call-outs:
N/A

### Testing:
Initial testing produced this data: https://app.codecov.io/github/aws/aws-lc/tree/codecov-ci/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: util/codecov-ci.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  if [ ! -e "$BUILD" ]; then
+    mkdir -p "$BUILD"
+  fi
+  BUILD=$(readlink -f "$BUILD")
+else
+  echo "Must specify a build directory."
+  exit 1
+fi
+
+cmake -DGCOV=1 -DDISABLE_PERL=1 -DBUILD_TESTING=1 -DBUILD_LIBSSL=1 -DCMAKE_BUILD_TYPE=Debug -S "${SRC}" -B "${BUILD}"
```
Comment (WillChilds-Klein) [inline util/codecov-ci.sh:29]: do we need debug symbols for coverage? should we consider `-DCMAKE_BUILD_TYPE=Release`? or `-DCMAKE_BUILD_TYPE=RelWithDebInfo`?

File: util/codecov-ci.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+cmake --build "${BUILD}" --target all_tests
+cmake --build "${BUILD}" --target run_tests
+
+#TODO: Use callgrind
```
Comment (WillChilds-Klein) [inline util/codecov-ci.sh:33]: what's blocking us from using this now?

File: .github/workflows/codecov-ci.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+concurrency:
+  group: code-cov-${{ github.workflow }}-${{ github.event.pull_request.number }}
+  cancel-in-progress: true
+jobs:
+  codecov-ci:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Install lcov
+        run: sudo apt-get -y install lcov
+      - uses: actions/checkout@master
```
Comment (WillChilds-Klein) [inline .github/workflows/codecov-ci.yml:16]: what's `master` here?

**Review by WillChilds-Klein — COMMENTED** [review]
very cool. the coverage graphs are excellent.

File: util/codecov-ci.sh (hunk 2 above)
Comment (justsmth) [reply util/codecov-ci.sh:33]: The PR provides test coverage data in our CI for C/C++ code that runs on Linux/x86-64, but it lacks data on any other combination of language, OS, and CPU. There are tools available (like Callgrind) that provide coverage for assembly code.  We should also extend this script to collect data for other CPU architectures.

But for now, I feel like this is a good "milestone" as this change quickly provides important test coverage information for code that is frequently updated (i.e., CPU/OS-agnostic C/C++ code).

File: .github/workflows/codecov-ci.yml (hunk 3 above)
Comment (justsmth) [reply .github/workflows/codecov-ci.yml:16]: I believe it's referring to a branch on the [actions/checkout repo](https://github.com/actions/checkout/tree/master). Good catch, I'll change it to [`v4`](https://github.com/actions/checkout/releases/tag/v4).

File: util/codecov-ci.sh (hunk 1 above)
Comment (justsmth) [reply util/codecov-ci.sh:29]: I chose `Debug` to avoid compiler optimizations that might obscure the coverage metrics.

**Review by skmcgrail — APPROVED** [review]

**Review by WillChilds-Klein — APPROVED** [review]

**Pull request merged**